File: results/2026-04-09_152435/tasks/11-semantic-version-bumper/typescript-bun-sonnet/generated-code/run-act-tests.ts
/**
 * Act Test Harness for Semantic Version Bumper
 *
 * This script:
 * 1. Creates a temp git repo with all project files
 * 2. Runs `act push --rm` to execute the GitHub Actions workflow in Docker
 * 3. Saves the full output to act-result.txt
 * 4. Asserts on exit code, job success, and exact version output values
 *
 * Run: bun run run-act-tests.ts
 */

import * as fs from "fs";
import * as path from "path";
import * as os from "os";

const CWD = process.cwd();
const ACT_RESULT_FILE = path.join(CWD, "act-result.txt");

// ─── Helpers ──────────────────────────────────────────────────────────────────

function log(msg: string): void {
  console.log(msg);
}

function copyDirRecursive(src: string, dst: string): void {
  if (!fs.existsSync(dst)) {
    fs.mkdirSync(dst, { recursive: true });
  }
  for (const entry of fs.readdirSync(src, { withFileTypes: true })) {
    // Skip node_modules and .git when copying project files
    if (entry.name === "node_modules" || entry.name === ".git") continue;
    const srcPath = path.join(src, entry.name);
    const dstPath = path.join(dst, entry.name);
    if (entry.isDirectory()) {
      copyDirRecursive(srcPath, dstPath);
    } else {
      fs.copyFileSync(srcPath, dstPath);
    }
  }
}

function runCommand(cmd: string[], cwd: string): { exitCode: number; output: string } {
  const proc = Bun.spawnSync(cmd, {
    cwd,
    stdout: "pipe",
    stderr: "pipe",
  });
  const stdout = proc.stdout ? new TextDecoder().decode(proc.stdout) : "";
  const stderr = proc.stderr ? new TextDecoder().decode(proc.stderr) : "";
  return {
    exitCode: proc.exitCode ?? 1,
    output: stdout + stderr,
  };
}

function assert(condition: boolean, message: string): boolean {
  if (condition) {
    log(`  ✓ ${message}`);
    return true;
  } else {
    log(`  ✗ FAIL: ${message}`);
    return false;
  }
}

// ─── Setup Temp Git Repo ──────────────────────────────────────────────────────

log("=== Setting up temp git repo ===");

const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "semver-act-"));
log(`Temp dir: ${tmpDir}`);

// Copy all project files into the temp directory
copyDirRecursive(CWD, tmpDir);

// Copy .actrc so act uses the right container image
const actrcSrc = path.join(CWD, ".actrc");
if (fs.existsSync(actrcSrc)) {
  fs.copyFileSync(actrcSrc, path.join(tmpDir, ".actrc"));
  log("Copied .actrc");
}

// Initialize git repo (act requires a git repo)
runCommand(["git", "init"], tmpDir);
runCommand(["git", "config", "user.email", "test@example.com"], tmpDir);
runCommand(["git", "config", "user.name", "Test Runner"], tmpDir);
runCommand(["git", "add", "-A"], tmpDir);
const commitResult = runCommand(["git", "commit", "-m", "ci: add semantic version bumper"], tmpDir);
log(`Git commit exit code: ${commitResult.exitCode}`);

// ─── Run act ──────────────────────────────────────────────────────────────────

log("\n=== Running act push --rm ===");
log("This takes 30-90 seconds for container startup + workflow execution...\n");

// --pull=false: use already-pulled container images, avoid slow pulls
// --rm: clean up container after run
const actResult = runCommand(["act", "push", "--rm", "--pull=false"], tmpDir);

const actOutput = actResult.output;
log(`act exit code: ${actResult.exitCode}`);

// ─── Save to act-result.txt ───────────────────────────────────────────────────

const delimiter = `\n${"=".repeat(80)}\n`;
const resultContent =
  `${delimiter}` +
  `ACT TEST RUN - Semantic Version Bumper\n` +
  `Timestamp: ${new Date().toISOString()}\n` +
  `Exit code: ${actResult.exitCode}\n` +
  `${delimiter}\n` +
  actOutput +
  `\n${delimiter}`;

fs.writeFileSync(ACT_RESULT_FILE, resultContent);
log(`\nAct output saved to: ${ACT_RESULT_FILE}`);

// Print act output for debugging
log("\n--- ACT OUTPUT ---");
log(actOutput.slice(0, 5000)); // First 5000 chars
if (actOutput.length > 5000) {
  log(`... (${actOutput.length - 5000} more chars, see act-result.txt)`);
}
log("--- END ACT OUTPUT ---\n");

// ─── Assertions ───────────────────────────────────────────────────────────────

log("=== Assertions ===");
let allPassed = true;

function check(condition: boolean, msg: string): void {
  if (!assert(condition, msg)) allPassed = false;
}

// 1. Act process exited with code 0
check(actResult.exitCode === 0, "act exited with code 0");

// 2. Job succeeded
check(actOutput.includes("Job succeeded"), 'Output contains "Job succeeded"');

// 3. Each test case passed with exact expected version
check(actOutput.includes("NEW_VERSION: 1.0.1"), "test-case-1: outputs exactly NEW_VERSION: 1.0.1");
check(actOutput.includes("NEW_VERSION: 1.1.0"), "test-case-2: outputs exactly NEW_VERSION: 1.1.0");
check(actOutput.includes("NEW_VERSION: 2.0.0"), "test-case-3: outputs exactly NEW_VERSION: 2.0.0");
check(actOutput.includes("NEW_VERSION: 1.3.0"), "test-case-4: outputs exactly NEW_VERSION: 1.3.0");

// 4. PASS markers present (workflow step-level pass messages)
check(actOutput.includes("PASS: test-case-1 patch bump"), "test-case-1: PASS marker present");
check(actOutput.includes("PASS: test-case-2 minor bump"), "test-case-2: PASS marker present");
check(actOutput.includes("PASS: test-case-3 major bump"), "test-case-3: PASS marker present");
check(actOutput.includes("PASS: test-case-4 mixed commits"), "test-case-4: PASS marker present");

// 5. Unit tests ran (bun test output)
check(actOutput.includes("pass") && actOutput.includes("bun test"), "Unit tests ran via bun test");

// 6. Version assertion in fixture verified (ASSERTION PASSED: X.Y.Z)
check(actOutput.includes("ASSERTION PASSED: 1.0.1"), "test-case-1: assertion passed");
check(actOutput.includes("ASSERTION PASSED: 1.1.0"), "test-case-2: assertion passed");
check(actOutput.includes("ASSERTION PASSED: 2.0.0"), "test-case-3: assertion passed");
check(actOutput.includes("ASSERTION PASSED: 1.3.0"), "test-case-4: assertion passed");

// ─── Cleanup ──────────────────────────────────────────────────────────────────

fs.rmSync(tmpDir, { recursive: true, force: true });
log(`\nCleaned up temp dir: ${tmpDir}`);

// ─── Final Result ─────────────────────────────────────────────────────────────

if (allPassed) {
  log("\n✓ ALL ACT TESTS PASSED");
  process.exit(0);
} else {
  log("\n✗ SOME ACT TESTS FAILED - check act-result.txt for details");
  process.exit(1);
}

File: results/2026-05-06_173435/tasks/13-dependency-license-checker/default-sonnet/generated-code/run_act_tests.py
"""
Act Integration Test Harness — Dependency License Checker

Sets up a temp git repo containing the project files, runs `act push --rm`
once, captures all output, saves it to act-result.txt, then asserts on
exact expected values in the output.

Limit: at most 3 `act push` runs total. We use one comprehensive run that
exercises all three test cases via the workflow steps.

Usage:
    python3 run_act_tests.py
"""
import os
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

# Where to write the captured act output
RESULT_FILE = Path(__file__).parent / "act-result.txt"

# Project root (the directory this script lives in)
PROJECT_ROOT = Path(__file__).parent.resolve()

# Files and directories to copy into the temp git repo
COPY_ITEMS = [
    "license_checker.py",
    "tests",
    "fixtures",
    ".github",
    ".actrc",
]


def run_command(cmd: list[str], cwd: str, timeout: int = 300) -> tuple[int, str]:
    """Run a command, capture combined stdout+stderr, return (exit_code, output)."""
    result = subprocess.run(
        cmd,
        cwd=cwd,
        capture_output=True,
        text=True,
        timeout=timeout,
    )
    return result.returncode, result.stdout + result.stderr


def setup_temp_repo(tmp_dir: str) -> None:
    """Copy project files into tmp_dir and initialise a git commit."""
    for item in COPY_ITEMS:
        src = PROJECT_ROOT / item
        if not src.exists():
            print(f"  Warning: {src} not found, skipping")
            continue
        dst = Path(tmp_dir) / item
        if src.is_dir():
            shutil.copytree(str(src), str(dst))
        else:
            shutil.copy2(str(src), str(dst))

    # Configure a minimal git identity so `git commit` works
    for cmd in [
        ["git", "init"],
        ["git", "config", "user.email", "test@example.com"],
        ["git", "config", "user.name", "Test Runner"],
        ["git", "add", "-A"],
        ["git", "commit", "-m", "test: add project files for act run"],
    ]:
        rc, out = run_command(cmd, cwd=tmp_dir)
        if rc != 0:
            raise RuntimeError(f"Setup command failed {cmd}: {out}")


def run_act(tmp_dir: str) -> tuple[int, str]:
    """Execute `act push --rm --pull=false` and return (exit_code, combined_output).

    --pull=false prevents act from trying to re-pull the local image from Docker Hub.
    """
    print("  Running: act push --rm --pull=false  (this takes 30-90 s) …")
    rc, out = run_command(
        ["act", "push", "--rm", "--pull=false"],
        cwd=tmp_dir,
        timeout=300,
    )
    return rc, out


def assert_exact(label: str, output: str, expected: str) -> None:
    """Assert that *expected* appears as a substring of *output*."""
    if expected not in output:
        print(f"  FAIL [{label}]: expected to find: {expected!r}")
        raise AssertionError(f"[{label}] Expected substring not found: {expected!r}")
    print(f"  PASS [{label}]: found {expected!r}")


def assert_not(label: str, output: str, unexpected: str) -> None:
    """Assert that *unexpected* does NOT appear in *output*."""
    if unexpected in output:
        print(f"  FAIL [{label}]: unexpected substring found: {unexpected!r}")
        raise AssertionError(f"[{label}] Unexpected substring found: {unexpected!r}")
    print(f"  PASS [{label}]: correctly absent: {unexpected!r}")


def run_test_cases() -> None:
    """
    Single act run covering all three test-case scenarios.

    Test case 1 — sample_package.json:
      express (MIT) → APPROVED, some-gpl-lib (GPL-3.0) → DENIED,
      unknown-pkg (not in DB) → UNKNOWN, Overall → NON-COMPLIANT

    Test case 2 — sample_requirements.txt:
      requests (Apache-2.0) → APPROVED, django (BSD-3-Clause) → APPROVED,
      some-gpl (GPL-2.0) → DENIED, Overall → NON-COMPLIANT

    Test case 3 — clean_package.json:
      express (MIT) → APPROVED, lodash (MIT) → APPROVED,
      Overall → COMPLIANT  (step must exit 0)
    """
    with tempfile.TemporaryDirectory(prefix="act_license_test_") as tmp_dir:
        print(f"\n[Setup] Temp repo: {tmp_dir}")
        setup_temp_repo(tmp_dir)

        print("[Act  ] Launching workflow …")
        exit_code, output = run_act(tmp_dir)

        # ----------------------------------------------------------------
        # Save the full output (required artifact)
        # ----------------------------------------------------------------
        delimiter = "=" * 70
        with open(RESULT_FILE, "w") as f:
            f.write(f"{delimiter}\n")
            f.write("ACT RUN: dependency-license-checker  (all three test cases)\n")
            f.write(f"{delimiter}\n")
            f.write(output)
            f.write(f"\n{delimiter}\n")
            f.write(f"Exit code: {exit_code}\n")
            f.write(f"{delimiter}\n")

        print(f"\n[Output] Saved to {RESULT_FILE}  ({len(output)} chars)")

        # ----------------------------------------------------------------
        # Assert: act exited with 0
        # ----------------------------------------------------------------
        if exit_code != 0:
            print("\n--- act output (last 80 lines) ---")
            for line in output.splitlines()[-80:]:
                print(line)
            raise AssertionError(
                f"act exited with code {exit_code} (expected 0)"
            )
        print(f"  PASS [act exit code]: 0")

        # ----------------------------------------------------------------
        # Assert: every job succeeded
        # ----------------------------------------------------------------
        assert_exact("job succeeded", output, "Job succeeded")

        # ----------------------------------------------------------------
        # Assert: pytest ran and all tests passed
        # ----------------------------------------------------------------
        assert_exact("pytest passed", output, "passed")
        # Ensure no test failures
        assert_not("pytest no failures", output, " failed")

        # ----------------------------------------------------------------
        # Test case 1 assertions — sample_package.json
        # ----------------------------------------------------------------
        # express@4.18.0 is MIT → APPROVED
        assert_exact("tc1 express approved", output, "express")
        assert_exact("tc1 APPROVED present", output, "APPROVED")

        # some-gpl-lib@1.0.0 is GPL-3.0 → DENIED
        assert_exact("tc1 some-gpl-lib present", output, "some-gpl-lib")
        assert_exact("tc1 DENIED present", output, "DENIED")

        # unknown-pkg not in DB → UNKNOWN
        assert_exact("tc1 unknown-pkg present", output, "unknown-pkg")
        assert_exact("tc1 UNKNOWN present", output, "UNKNOWN")

        # Overall NON-COMPLIANT for sample_package.json
        assert_exact("tc1 NON-COMPLIANT", output, "NON-COMPLIANT")

        # ----------------------------------------------------------------
        # Test case 2 assertions — sample_requirements.txt
        # ----------------------------------------------------------------
        # requests Apache-2.0 → APPROVED
        assert_exact("tc2 requests present", output, "requests")
        # some-gpl GPL-2.0 → DENIED
        assert_exact("tc2 some-gpl present", output, "some-gpl")

        # ----------------------------------------------------------------
        # Test case 3 assertions — clean_package.json (all approved)
        # ----------------------------------------------------------------
        # lodash MIT → APPROVED (only in clean fixture)
        assert_exact("tc3 lodash present", output, "lodash")
        # Overall COMPLIANT (not NON-COMPLIANT) for the clean run
        # NON-COMPLIANT already asserted above; just verify COMPLIANT appears
        assert_exact("tc3 COMPLIANT", output, "COMPLIANT")

        # ----------------------------------------------------------------
        # Exact summary counts (from the compliance report header lines)
        # ----------------------------------------------------------------
        # sample_package.json has 5 deps: express MIT, react MIT, jest MIT,
        # some-gpl-lib GPL-3.0 (denied), unknown-pkg (unknown)
        # → "3 approved, 1 denied, 1 unknown"
        assert_exact(
            "tc1 summary counts",
            output,
            "3 approved, 1 denied, 1 unknown",
        )

        # sample_requirements.txt: requests Apache-2.0 (approved),
        # django BSD-3-Clause (approved), some-gpl GPL-2.0 (denied),
        # no-version-pkg (unknown)
        # → "2 approved, 1 denied, 1 unknown"
        assert_exact(
            "tc2 summary counts",
            output,
            "2 approved, 1 denied, 1 unknown",
        )

        # clean_package.json: express MIT (approved), lodash MIT (approved)
        # → "2 approved, 0 denied, 0 unknown"
        assert_exact(
            "tc3 summary counts",
            output,
            "2 approved, 0 denied, 0 unknown",
        )

    print("\n[All assertions passed]")


def main() -> int:
    print("=" * 60)
    print("Dependency License Checker — Act Integration Tests")
    print("=" * 60)
    try:
        run_test_cases()
        print("\nResult: SUCCESS")
        return 0
    except AssertionError as exc:
        print(f"\nResult: FAILURE — {exc}")
        return 1
    except Exception as exc:
        print(f"\nResult: ERROR — {exc}")
        return 1


if __name__ == "__main__":
    sys.exit(main())
